### PR TITLE
Show aggregate users under groups

### PR DESCRIPTION
**Pull request by wss-rbunning**
## Hey, I just made a Pull Request!

This PR adds a toggle switch to the members card, allowing you to toggle between showing direct members and aggregated members of a group. Aggregated users is recursive and handles duplicated users and child groups. 

This addresses feature request #16959 

## Before:
image

## After:
image
image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by ahhhndre** [comment]
@wss-rbunning, is this opt in? As in we won’t see the switch at all if we don’t enable it.

**Comment by wss-rbunning** [comment]
@ahhhndre As it is right now, the switch is not opt-in and will always be visible. It does, however, default to the off position when the page loads, showing only direct members like before.

**Comment by awanlin** [comment]
> @ahhhndre As it is right now, the switch is not opt-in and will always be visible. It does, however, default to the off position when the page loads, showing only direct members like before.

The general practice is to make these types of things Opt-In. Doing it this way people can choose to have it visible versus it being force upon them. Should be a pretty easy thing to implement. Can we please change that?

(Sorry if this is confusing, I am Ahhhndre I just wanted to follow up on this when I saw it on my phone before I forgot and it got merged in)

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for this contribution @wss-rbunning, please don't take my comments about making this opt-in to seem like it's not wanted. I know that this is something the Community will be happy to have once it's merged in and released 🚀 

Overall this looks good to me, left a few comments

File: plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx [hunk 1]
```diff
@@ -193,6 +218,16 @@ export const MembersListCard = (props: {
         subheader={`of ${displayName}`}
         {...(nbPages <= 1 ? {} : { actions: pagination })}
       >
+        Direct Members
+        <Switch
```
[the remark this answers is not in the document]
Comment (wss-rbunning) [reply plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx]: I have made the aggregate users toggle switch be opt-in now. There is now a showAggregateMembersToggle on MembersListCard that when set will show the toggle. If it is not set, the card will not look different.

File: plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx
[the remark this answers is not in the document]
Comment (wss-rbunning) [reply plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx:1]: I have created a Storybook entry for when the toggle switch is turned on. From the Storybook page, you can toggle the switch on and off to see the aggregated users get added or removed.

File: .changeset/new-roses-fail.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-org': patch
+---
+
+Changed the MembersListCard component to allow displaying aggregated members when viewing a group. Now, a toggle switch is displayed that lets you switch between showing direct members and aggregated members.
```
[the remark this answers is not in the document]
Comment (wss-rbunning) [reply .changeset/new-roses-fail.md]: I have now added a small code snippet to the change set that demonstrates how to enable the aggregate users toggle switch.

**Comment by wss-rbunning** [comment]
@awanlin I fully understand the reason for wanting to have it be opt-in and I appreciate the feedback. I have pushed changes that should address your feedback items. Please let me know if there is any other feedback or suggestions you would like me to address 👍.

File: plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx:1]: Thanks! 👍

**Review by awanlin — APPROVED** [review]
LGTM! 🚀

File: plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx]: Thanks! 👍

File: .changeset/new-roses-fail.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (awanlin) [reply .changeset/new-roses-fail.md]: Thanks! 👍

**Comment by wss-rbunning** [comment]
@awanlin: 
@jtreher brought this to my attention: With my current implementation of the switch it loads the aggregated users regardless of whether or not the switch is requested through the prop or flipped on. 

With how it is now, it loads the aggregated users in the background, not delaying the page load while this happens. If the switch is flipped to "Aggregate Users" before all aggregate users are done loading, no additional users will be displayed until it is finished.

I am going to change this so that it will only run the logic to get aggregated users if the "show the switch" prop is set (so it isn't doing a lot of extra stuff that someone isn't asking for). Do we want to show a loading bar if the switch is flipped and aggregate users are loading?

**Comment by awanlin** [comment]
Sounds good to me @wss-rbunning, I think showing a progress bar would be good. How do they deal with that for the Ownership Card? They do something very similar. We don't use it though so I don't remember if it has a progress bar or not.

**Comment by wss-rbunning** [comment]
@awanlin It looks like in the Ownership Card they have a separate "ownership grid" component that is a child of the card itself and does not contain the toggle switch. This grid shows a loading bar when it is loading members. 

I have pushed changes with my suggestions. Here is the new behavior:

- When the toggle switch is not enabled, the MembersCard will not bother to load descendant members.
- The existing overall loading bar displayed when loading members on the MemebersCard is still used. It is, however, only used when loading direct members. The loading of descendant members is still asynchronous and does not delay rendering the page or direct members. If the user does flip the switch over to aggregate members before it is done enumerating descendant members, a loading bar will be displayed in place of members until it is finished.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you, looks good! 👍 

Just a couple of nits

File: plugins/org/package.json [hunk 3]
```diff
@@ -28,6 +28,7 @@
     "clean": "backstage-cli package clean"
   },
   "dependencies": {
+    "@backstage/catalog-client": "workspace:^",
```
[the remark this answers is not in the document]
Comment (wss-rbunning) [reply plugins/org/package.json]: I have moved the testing helpers to a ```__testUtils__``` folder and moved this dependency to being a dev dependency 👍. I am only using these mocks in the tests and storybook.

File: plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx [hunk 4]
```diff
@@ -165,6 +189,18 @@ export const MembersListCard = (props: {
...
 
+  const members = removeDuplicateEntitiesFrom(
+    [
+      ...(directMembers ?? []),
+      ...(descendantMembers && showAggregateMembers ? descendantMembers : []),
+    ].sort((a, b) => {
+      const nameToCompareInA = a.spec.profile?.displayName ?? a.metadata.name;
+      const nameToCompareInB = b.spec.profile?.displayName ?? b.metadata.name;
+
+      return nameToCompareInA.localeCompare(nameToCompareInB);
```
[the remark this answers is not in the document]
Comment (wss-rbunning) [reply plugins/org/src/components/Cards/Group/MembersList/MembersListCard.tsx]: I have switched this to now compare based off of the results of stringifyEntityRef 👍

File: plugins/org/src/components/Cards/Group/MembersList/MembersListCard.stories.tsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021 The Backstage Authors
+ * Copyright 2020 The Backstage Authors
```
Comment (Rugvip) [inline plugins/org/src/components/Cards/Group/MembersList/MembersListCard.stories.tsx:2]:
```suggestion
 * Copyright 2021 The Backstage Authors
```

Oops, bit too far back here 😅

**Review by Rugvip — APPROVED** [review]
Nice, thank you! 😁 

Just a tiny think that I'll tweak and then good to :shipit:

**Pull request merged**